src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/SessionParameterValueResolver.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\MapSessionParameter;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+
+final class SessionParameterValueResolver implements ValueResolverInterface
+{
+    public function resolve(Request $request, ArgumentMetadata $argument): array
+    {
+        if (!$request->hasSession()) {
+            return [];
+        }
+
+        if (!$argument->getAttributesOfType(MapSessionParameter::class, ArgumentMetadata::IS_INSTANCEOF)) {
+            return [];
+        }
+
+        if ((!$type = $argument->getType()) || !class_exists($type)) {
+            throw new \LogicException(sprintf('#[MapSessionParameter] cannot be used on controller argument "$%s": "%s" is not a class name.', $argument->getName(), $type));
+        }
+
+        if ($request->getSession()->has($type)) {
+            return [
+                $request->getSession()->get($type),
+            ];
+        }
+
+        $object = ($argument->hasDefaultValue() && $argument->getDefaultValue()) ? $argument->getDefaultValue() : new $type();
```

Comment:
That's a good idea, but I would always respect the default value - it's there for a reason:
```suggestion
        if (\is_object($object = $argument->hasDefaultValue() ? $argument->getDefaultValue() : new $type()) {
            $request->getSession()->set($name, $object);
        }
```

Then yes, a non-object value won't be linked to the session. But you can still to it manually after creating the default object:
```php
$foo ?? $session->set('foo', $foo = new Foo([...]));
```